src/shared/discover-image.c

```diff
@@ -349,23 +349,36 @@ static int image_make(
          * This function returns -ENOENT if we can't find the image after all, and -EMEDIUMTYPE if it's not a file we
          * recognize. */
 
+        _cleanup_close_ int _fd = -EBADF;
+        if (fd < 0) {
+                /* If we didn't get an fd passed in, then let's pin it via O_PATH now */
+                _fd = openat(dir_fd, filename, O_PATH|O_CLOEXEC);
+                if (_fd < 0)
+                        return -errno;
+
+                fd = _fd;
+                st = NULL; /* refresh stat() data now that we have the inode pinned */
+        }
+
+        struct stat stbuf;
         if (!st) {
-                if (fstatat(dfd, filename, &stbuf, 0) < 0)
+                if (fstat(fd, &stbuf) < 0)
                         return -errno;
 
                 st = &stbuf;
         }
 
-        if (!path)
-                (void) fd_get_path(dfd, &parent);
+        _cleanup_free_ char *parent = NULL;
+        if (!dir_path) {
+                (void) fd_get_path(dir_fd, &parent);
+                dir_path = parent;
+        }
 
         read_only =
-                (path && path_startswith(path, "/usr")) ||
-                (parent && path_startswith(parent, "/usr")) ||
```

Comment:
parent check is lost, it was recently added, is that intentional?

Reply:
yes, the patch merges the two vars into one, which is nicer and simpler.